src/persist-client/src/internal/state_fetcher.rs

```diff
@@ -0,0 +1,373 @@
+        assert!(
+            hint.first().map_or(true, |x| x.seqno > state.seqno),
+            "update_state {} called with invalid hint {:?}",
+            state.seqno,
+            hint.iter().map(|x| x.seqno).collect::<Vec<_>>()
+        );
+
+        // If we have at least one live diff that can advance state, apply it
+        // and return.
+        match hint.first() {
+            Some(x) if x.seqno == state.seqno.next() => {
+                self.metrics.state.update_state_fast_path.inc();
+                self.apply_diffs(state, hint.iter());
+                return Ok(());
+            }
+            _ => {
+                self.metrics.state.update_state_slow_path.inc();
+                // Fall through.
+            }
+        }
+        debug!(
+            "update_state didn't hit update_state fast path {} {:?}",
+            state.seqno,
+            hint.iter().map(|x| x.seqno).collect::<Vec<_>>()
+        );
+
+        // Otherwise, we've gc'd the diffs we'd need to advance state to where
+        // hint is at. Use the hint to kick off fetching a new rollup and use
+        // the result of that.
+        self.metrics.state.update_state_slow_path.inc();
+        *state = self.fetch_latest_state(&state.shard_id, hint).await?;
+        Ok(())
+    }
+
+    pub fn apply_diffs<'a, I: IntoIterator<Item = &'a VersionedData>>(
+        &self,
+        state: &mut State<K, V, T, D>,
+        diffs: I,
+    ) {
+        let mut state_seqno = state.seqno;
+        let diffs = diffs.into_iter().filter_map(move |diff| {
+            if diff.seqno == state_seqno {
+                // No-op.
+                return None;
+            }
+            let diff = self.decode_diff(diff);
+            assert_eq!(diff.seqno_from, state_seqno);
+            state_seqno = diff.seqno_to;
+            Some(diff)
+        });
+        state
+            .apply_diffs(&self.metrics, diffs)
+            .expect("state diff should apply cleanly");
+    }
+
+    // WIP Returns all diffs and a rollup at the version == the first diff
+    pub async fn fetch_live_states(
+        &self,
+        shard_id: &ShardId,
+    ) -> Result<(State<K, V, T, D>, Vec<VersionedData>), CodecMismatch> {
+        let retry = self
+            .metrics
+            .retries
+            .fetch_live_states
+            .stream(Retry::persist_defaults(SystemTime::now()).into_retry_stream());
+        loop {
+            let live_diffs = self.fetch_live_diffs(&shard_id).await;
+            let earliest_live_diff = match live_diffs.first() {
+                Some(x) => x,
+                None => {
+                    // Live diffs is only empty if the shard is uninitialized.
+                    // Initialize it and try again. This throws away the return
+                    // value of maybe_init, but that's fine because we should
+                    // never hit this case in practice.
+                    let _ = self.maybe_init(shard_id).await?;
```

Comment:
could potentially avoid the are-we-init-or-not questions by creating two Fetcher types. calling `maybe_init` on the first returns a Fetcher that can assume its shard is always initialized. IMO this isn't total bikeshedding, I think it could clarify the logic in here a bit

Reply:
I think this one is just a confusing comment, i've replaced it with a panic, which is what I should have done in the first place. the only method in this whole file that even considers an initialized shard, is maybe_init_shard (and now try_compare_and_set_current)

I _think_ we could also make Fetcher::new into something that returns `(Self, State<K, V, T, D>)` and then Machine::new would take State as an argument (guaranteeing that Machine is only ever dealing with initialized shard) but I haven't gotten there yet

Reply:
> I think we could also make Fetcher::new into something that returns (Self, State<K, V, T, D>) and then Machine::new would take State as an argument (guaranteeing that Machine is only ever dealing with initialized shard) but I haven't gotten there yet

played with this for a bit, but couldn't manage anything that didn't feel like a significant readability regression to me. I added an explicit mention in the rustdoc for each method whether it can be called on an uninitialized shard or not, maybe that helps?